### ISSUE TITLE
Bump Go to 1.16 in post-cluster-api-push-images

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#4383]
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

I saw because of a failed [post-cluster-api-provider-openstack-push-images](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-cluster-api-provider-openstack-push-images/1375280030887710720) job in CAPO that the images we're using in `cloudbuild.yaml` are outdated. The CAPO image was even more outdated as in other repos, that's why the job failed in our repo but not in others.

This affects all providers as far as I could see and the latest `gcr.io/k8s-testimages/gcb-docker-gcloud` image available is using Go 1.13:
https://github.com/kubernetes/test-infra/blob/166e9a2c36b8645ae7ddfdfdfc9671c214c61e77/images/gcb-docker-gcloud/cloudbuild.yaml#L12

**What did you expect to happen:**

I would have expected we're using the Go version we're using everywhere else (1.16) in these jobs.

**Anything else you would like to add:**

This might be intentional as we have the same Prow job configuration for all our branches including old release branches.

If we want to bump the go version, I would open the necessary PRs in test-infra and in the affected ClusterAPI repositories.

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by fabriziopandini** [comment]
@sbueringer I'm generally +1 for keeping up with go versions, but I'm trying to better understand the problem.

It seems that CAPI images are published (the last one is https://console.cloud.google.com/gcr/images/k8s-staging-cluster-api/GLOBAL/capd-manager@sha256:613fded939b84c6716adbcbbaf7e30220a05ad39dbf81b7a938c0b6acfe08611/details?tab=info&project=k8s-staging-cluster-api), so I'm not sure that having go version in sync here is a real requirement.

Also, for the changes in test infra I would suggest to ask on sig-testing, given that AFAIK the image promotion process is shared across many projects.

**Comment by sbueringer** [comment]
@fabriziopandini 

It's not a problem which leads to errors at the moment anywhere (we were a bit more behind in CAPO).
I took another look and realized that the images we publish are build with docker multi-stage build anyway, 
so they are not affected.

So this is  could only lead to problems in cases where we depend on the local go version in our Makefiles.
As far as I know we're only compiling our hack/tools go binaries with local Go.

tl;dr it's mostly a matter if we want to keep that in sync with what we're using locally (I assume most of us are using Go 1.16)
It only will break if one of our hack/tools depend on a newer Go version. (this was the case between Go 1.12 and Go 1.13)

**Comment by vincepri** [comment]
We should definitely try to keep the Go version in sync with the target Kubernetes version we're using. For now v0.3.x is on Go 1.13, while v0.4.x is going to use Go 1.16 and Kubernetes 1.21
